src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -381,47 +391,70 @@ DropShards(Oid relationId, char *schemaName, char *relationName,
 		CoordinatedTransactionUse2PC();
 	}
 
-	ShardInterval *shardInterval = NULL;
-	foreach_ptr(shardInterval, deletableShardIntervalList)
+	List *dropTaskList = DropTaskList(relationId, schemaName, relationName,
+									  deletableShardIntervalList);
+
+	Task *task = NULL;
+	foreach_ptr(task, dropTaskList)
 	{
-		uint64 shardId = shardInterval->shardId;
+		uint64 shardId = task->anchorShardId;
 		char *shardRelationName = pstrdup(relationName);
 
-		Assert(shardInterval->relationId == relationId);
-
 		/* build shard relation name */
 		AppendShardIdToName(&shardRelationName, shardId);
 
-		List *shardPlacementList = ShardPlacementList(shardId);
-
 		ShardPlacement *shardPlacement = NULL;
-		foreach_ptr(shardPlacement, shardPlacementList)
+		foreach_ptr(shardPlacement, task->taskPlacementList)
 		{
 			uint64 shardPlacementId = shardPlacement->placementId;
+			int32 shardPlacementGroupId = shardPlacement->groupId;
 
-			if (shardPlacement->groupId == COORDINATOR_GROUP_ID &&
-				IsCoordinator() &&
+			if (shardPlacementGroupId == COORDINATOR_GROUP_ID && IsCoordinator() &&
 				DropSchemaOrDBInProgress())
 			{
 				/*
 				 * The active DROP SCHEMA/DATABASE ... CASCADE will drop the
 				 * shard, if we try to drop it over another connection, we will
-				 * get into a distributed deadlock.
+				 * get into a distributed deadlock. Hence, just delete the shard
+				 * placement metadata and skip it for now.
 				 */
+				DeleteShardPlacementRow(shardPlacementId);
+				continue;
 			}
-			else
-			{
-				char storageType = shardInterval->storageType;
 
-				const char *dropShardPlacementCommand =
-					CreateDropShardPlacementCommand(schemaName, shardRelationName,
-													storageType);
+			bool isLocalShardPlacement = (shardPlacementGroupId == localGroupId);
+
+			/*
+			 * If it is a local placement of a distributed table, then try to
+			 * execute the DROP command locally.
+			 */
+			if (isLocalShardPlacement && ShouldExecuteTasksLocally(dropTaskList))
```

Comment:
`ShouldExecuteTasksLocally(dropTaskList)` is not free. Can't we calculate once and use it in the loop?